autospec/download.py

```diff
@@ -0,0 +1,71 @@
+#!/usr/bin/true
+#
+# download.py - part of autospec
+# Copyright (C) 2018 Intel Corporation
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+from io import BytesIO
+import os
+import sys
+
+import pycurl
+from util import print_fatal
+
+
+def do_curl(url, dest=None, post=None, is_fatal=False):
+    """Perform a curl operation for URL.
+
+    If POST is set, a POST is performed with fields taken from the specified
+    value. Otherwise a GET is performed.  If DEST is set, the curl response (if
+    successful) is written to the specified path and the path is returned.
+    Otherwise the response is returned as a BytesIO object. If any errors
+    occur, None is returned.
+    """
+    c = pycurl.Curl()
+    c.setopt(c.URL, url)
+    if post:
+        c.setopt(c.POSTFIELDS, post)
+    c.setopt(c.FOLLOWLOCATION, True)
+    c.setopt(c.FAILONERROR, True)
+    buf = BytesIO()
+    c.setopt(c.WRITEDATA, buf)
+    try:
+        c.perform()
+    except pycurl.error as e:
+        if is_fatal:
+            print_fatal("Unable to fetch {}: {}".format(url, e))
+            sys.exit(1)
+        return None
+    finally:
+        c.close()
+
+    # write to dest if specified
+    if dest:
+        try:
+            with open(dest, 'wb') as fp:
+                fp.write(buf.getvalue())
+        except Exception as e:
+            if os.path.exists(dest):
```

Comment:
I think I'm inclined to leave cleanup of the file to open rather than handle unlinks in the exception handler.

Reply:
If an exception is raised, I don't see the file unlinked though... I left it out originally

Reply:
Hrm well my worry is that the file can't be opened for writing due to it being opened by another process and this will cause another exception. Maybe an exception for IOError that does the unlink and a generic exception that doesn't? unlink is annoying in that there isn't an option for it to ignore failures.

Reply:
Perhaps don't catch the generic exception even.

Reply:
Okay, I will just handle `IOError`

Reply:
I made this change in the latest commit